### DISCUSSION TITLE
ESP32  NodeMCU D1 R32 board and 2.4" ILI9341Touch Display compile error

### DISCUSSION THREAD

**Discussion by karstengit** [Bodmer/TFT_eSPI#2299] [Compatible displays and setup files]
I am trying to use this display:
http://www.lcdwiki.com/res/MAR2406/2.4inch_Arduino_8BIT_Module_MAR2406_User_Manual_EN.pdf

on such an ESP32 board:
https://www.az-delivery.de/blogs/azdelivery-blog-fur-arduino-und-raspberry-pi/d1-r32

The reason is that the display can simply be plugged on this board.

User_Setup.h is:
```
// Only define one driver, the other ones must be commented out
#define ILI9341_DRIVER

// Tell the library to use 8 bit parallel mode (otherwise SPI is assumed)
#define TFT_PARALLEL_8_BIT

// The ESP32 and TFT the pins used for testing are:
#define TFT_CS   33  // Chip select control pin (library pulls permanently low
#define TFT_DC   15  // Data Command control pin - must use a pin in the range 0-31
#define TFT_RST  32  // Reset pin, toggles on startup

#define TFT_WR    4  // Write strobe control pin - must use a pin in the range 0-31
#define TFT_RD    2  // Read strobe control pin

#define TFT_D0   12  // Must use pins in the range 0-31 for the data bus
#define TFT_D1   13  // so a single register write sets/clears all bits.
#define TFT_D2   26  // Pins can be randomly assigned, this does not affect
#define TFT_D3   25  // TFT screen update performance.
#define TFT_D4   17
#define TFT_D5   16
#define TFT_D6   27
#define TFT_D7   14

```

with this TFT Test:

```
#include "SPI.h"
#include "TFT_eSPI.h"

// Use hardware SPI
TFT_eSPI tft = TFT_eSPI();

unsigned long total = 0;
unsigned long tn = 0;

void setup() {
  Serial.begin(115200);
  while (!Serial);
  Serial.println(""); Serial.println("");
  Serial.println("Bodmer's TFT_eSPI library Test!"); 
 
  tft.init();
}
...
```
Trying to compile ends with

```
In file included from /.../arduino15/packages/esp32/hardware/esp32/2.0.3/cores/esp32/Arduino.h:36,
                 from /.../Arduino/libraries/TFT_eSPI/TFT_eSPI.h:30,
                 from /.../Arduino/libraries/TFT_eSPI/TFT_eSPI.cpp:17:
/.../Arduino/libraries/TFT_eSPI/Processors/TFT_eSPI_ESP32.c: In member function 'void TFT_eSPI::gpioMode(uint8_t, uint8_t)':
/.../Arduino/libraries/TFT_eSPI/Processors/TFT_eSPI_ESP32.c:132:32: error: 'esp32_gpioMux' was not declared in this scope
   ESP_REG(DR_REG_IO_MUX_BASE + esp32_gpioMux[gpio].reg) // Register lookup
                                ^~~~~~~~~~~~~
/.../arduino15/packages/esp32/hardware/esp32/2.0.3/cores/esp32/esp32-hal.h:73:47: note: in definition of macro 'ESP_REG'
 #define ESP_REG(addr) *((volatile uint32_t *)(addr))
                                               ^~~~

exit status 1
Fehler beim Kompilieren für das Board ESP32vn IoT Uno.
```

Any idea how to get this error **'esp32_gpioMux' was not declared** solved?

Maybe it depends on the Arduino-Environment?
https://github.com/espressif/arduino-esp32/issues/6559

**Comment by karstengit** [comment]
The settings for the board are correct.
A sketch for scanning Bluetooth devices compiles and is working correct with output on the UART.

The first sketch compiles when the User_Setup.h is configured for an SPI-Interface.
So this is clear a case of **TFT_PARALLEL_8_BIT**

**Comment by Bodmer** [comment]
This error will occur if old versions of TFT_eSPI (> 1 year old) are used with recent ESP32 board packages >= 2.0.0. The error occurs because Esspressif made extensive changes to the ESP32 board package at 2.0.0 and removed or renamed some low level functions.

You have 2 options:
1. Upgrade to a new version of TFT_eSPI
2. Downgrade to an older than 2.0.0 version of the board package

**Reply by karstengit** [reply]
I beg your pardon - it is my fault not to update the library first.

Now it compiles, but the display is not working in my first test.
When I have found out the reason I will post again.

**Reply by bionicbone** [reply]
> This error will occur if old versions of TFT_eSPI (> 1 year old) are used with recent ESP32 board packages >= 2.0.0. The error occurs because Esspressif made extensive changes to the ESP32 board package at 2.0.0 and removed or renamed some low level functions.
> 
> You have 2 options:
> 
> 1. Upgrade to a new version of TFT_eSPI
> 2. Downgrade to an older than 2.0.0 version of the board package

Thanks Bodmer for a fantastic library and the perfect support. 

I reinstalled my workstation today, and although I installed the library via VS IDE for some reason it installed the old version 2.3.x ,I found my issue with a quick google search and now see the current version available is 2.5.43 in VS IDE.  It compiles now, thanks again.

**Comment by karstengit** [comment]
Now the Pins have been compared and a complete new pin assignment has been tested.

![LCD](https://user-images.githubusercontent.com/5941550/211210224-37f47165-cbd6-4fc5-86e5-4a01f5ccbce8.jpg)

![Pinout_Arduino_Uno](https://user-images.githubusercontent.com/5941550/211209597-c139263c-bf54-4b04-a774-0e9567ffe06d.jpg)

![Pinout_ESP32_D1_R32](https://user-images.githubusercontent.com/5941550/211209598-15ce34b6-5dc3-497a-abe4-6ba6a3ddc23b.jpg)

```
// Tell the library to use 8 bit parallel mode (otherwise SPI is assumed)
#define TFT_PARALLEL_8_BIT

// The ESP32 and TFT the pins used for testing are:
#define TFT_CS   17  // Chip select control pin (library pulls permanently low
#define TFT_DC   16  // Data Command control pin - must use a pin in the range 0-31
#define TFT_RST  18  // Reset pin, toggles on startup

#define TFT_WR   15  // Write strobe control pin - must use a pin in the range 0-31
#define TFT_RD   14  // Read strobe control pin

#define TFT_D0    8  // Must use pins in the range 0-31 for the data bus
#define TFT_D1    9  // so a single register write sets/clears all bits.
#define TFT_D2    2  // Pins can be randomly assigned, this does not affect
#define TFT_D3    3  // TFT screen update performance.
#define TFT_D4    4
#define TFT_D5    5
#define TFT_D6    6
#define TFT_D7    7
```

This compiles and has been flashed, but result in an boot loop at the point calling **tft.init()** :

```
Bodmer's TFT_eSPI library Test!
ets Jun  8 2016 00:22:57

rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:12812
load:0x40080400,len:3032
entry 0x400805e4


Bodmer's TFT_eSPI library Test!
ets Jun  8 2016 00:22:57

...
```

```
Creating esp32 image...
Merged 2 ELF sections
Successfully created esp32 image.
python3 /.../arduino15/packages/esp32/hardware/esp32/2.0.3/tools/gen_esp32part.py -q /tmp/arduino_build_69137/partitions.csv /tmp/arduino_build_69137/ESP32-SEM6000.ino.partitions.bin
Bibliothek SPI in Version 2.0.0 im Ordner: /.../arduino15/packages/esp32/hardware/esp32/2.0.3/libraries/SPI  wird verwendet
Bibliothek TFT_eSPI in Version 2.4.79 im Ordner: /.../libraries/TFT_eSPI  wird verwendet
Bibliothek FS in Version 2.0.0 im Ordner: /.../arduino15/packages/esp32/hardware/esp32/2.0.3/libraries/FS  wird verwendet
Bibliothek SPIFFS in Version 2.0.0 im Ordner: /.../arduino15/packages/esp32/hardware/esp32/2.0.3/libraries/SPIFFS  wird verwendet
/.../.arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/gcc8_4_0-esp-2021r2-patch3/bin/xtensa-esp32-elf-size -A /tmp/arduino_build_69137/ESP32-SEM6000.ino.elf
Der Sketch verwendet 282089 Bytes (21%) des Programmspeicherplatzes. Das Maximum sind 1310720 Bytes.
Globale Variablen verwenden 17676 Bytes (5%) des dynamischen Speichers, 310004 Bytes für lokale Variablen verbleiben. Das Maximum sind 327680 Bytes.

```

At this moment there is no idea why?

**Reply by Bodmer** [reply]
The GPIO numbers must be used, so for TFT D0 the correct allocation is:
#define TFT_D0    12

Since some of the GPIO are input only wires need to be added and the correct output pin used. See ReadMe here which includes a link to an example setup file:
https://github.com/Bodmer/TFT_eSPI#8-bit-parallel-support

**Reply by karstengit** [reply]
Thank you and sorry again that I did not found / read this.

**Reply by Bodmer** [reply]
No problem! If you need more help just ask.
